packages/repository/src/repository-mixin.ts

```diff
@@ -0,0 +1,127 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import { Class } from './common-types';
+import { MixinBuilder } from './mixin';
+import { Component } from '@loopback/core';
+import { Repository } from './repository';
+
+// tslint:disable:no-any
+
+/**
+* A mixin class for Application that creates a .repository()
+* function to register a repository automatically. Also overrides
+* component function to allow it to register repositories automatically.
+*
+* ```ts
+*
+* class MyApplication extends RepositoryMixin(Application) {}
+* ```
+*/
+export function RepositoryMixin<T extends Class<any>>(superClass: T) {
+  return class extends superClass {
+    // A mixin class has to take in a type any[] argument!
+    constructor(...args: any[]) {
+      super(...args);
+      if (!this.options) this.options = {};
+
+      if (this.options.repositories) {
+        for (const repo of this.options.repositories) {
+          this.repository(repo);
+        }
+      }
+
+      if (this.options.components) {
+        // Super would have already mounted the component
+        for (const component of this.options.components) {
+          this.mountComponentRepository(component);
+        }
+      }
+    }
+
+    /**
+    * Add a repository to this application.
+    *
+    * @param repo The repository to add.
+    *
+    * ```ts
+    *
+    * class NoteRepo {
+    *   model: any;
+        *
+    *   constructor() {
+    *     const ds: juggler.DataSource = new DataSourceConstructor({
+    *       name: 'db',
+    *       connector: 'memory',
+    *     });
+    *
+    *     this.model = ds.createModel(
+    *       'note',
+    *       {title: 'string', content: 'string'},
+    *       {}
+    *     );
+    *   }
+    * };
+    *
+    * app.repository(NoteRepo);
+    * ```
+    */
+    repository(repo: Class<Repository<any>>) {
+      const repoKey = `repositories.${repo.name}`;
+      this.bind(repoKey).toClass(repo);
+    }
+
+    /**
+    * Add a component to this application. Also mounts
+    * all the components repositories.
+    *
+    * @param component The component to add.
+    *
+    * ```ts
+    *
+    * export class ProductComponent {
+    *   controllers = [ProductController];
+    *   repositories = [ProductRepo, UserRepo];
+    *   providers = {
+    *     [AUTHENTICATION_STRATEGY]: AuthStrategy,
+    *     [AUTHORIZATION_ROLE]: Role,
+    *   };
+    * };
+    *
+    * app.component(ProductComponent);
+    * ```
+    */
+    public component(component: Component) {
```

Comment:
The mixin overrides the `component()` method so that it can inject logic to handle `repositories` from a component. 

There is another way to use `extension points/extension` to support such cases. For example, the `repository` component can contribute a binding for component.repositories to the context. In the application.component() implementation, it can find registered bindings for extended artifacts for a component, in this case, the `repositories` will be delegated to the `repository` extension.

Reply:
Just created this issue https://github.com/strongloop/loopback-next/issues/543 as per our discussion regarding your comment.